### PR TITLE
channeldb: index payments by payment addr, use payment hash as fallback

### PR DESCRIPTION
**Pull request by cfromknecht**
This PR adds a new persistent index on invoices to allow efficient
lookups and updates by payment address. This is stepping stone
to enable AMP payments where the payment hash will not be
associated with the target invoice.

Since 0.9, all new invoices include a random payment address,
however older invoices do not have this value. New invoices will
be added to the payment address index on a rolling basis, but 
the payment hash index will continue to serve as a fallback in case
the target invoice was created prior to this commit. As such,
no migration is needed for invoices created prior to this commit, 
but we include one to bump the db version number and initialize
the top-level index bucket.

This also unblocks #4167, since it allows us to insert a random
payment address for experimental keysend payments. In the future,
where keysend is rolled into the final AMP proposal, this ensures that 
all keysend (experimental or otherwise) have a payment address
associated with the invoice. This is desired since we will be using
payment addresses entirely to identify and update AMP payments, so
we ensure the code paths between AMP and legacy keysend are
identical.

Further off, once payment addresses are required, we can stop inserting
invoices into the payment hash index entirely. Currently however, we rely
on invoices being indexed to support other operations, e.g. hodl settle/cancel
or legacy senders that don't understand MPP.

### CODE REVIEW

**Comment by joostjager** [comment]
One thought that came up: (hodl) keysend payments and AMP payments can be settled by payment address. But how does that work for recurring payments to the same address? How to keep those apart when you want to settle or cancel? I think the `set_id` is needed? But does that also require another index?

File: channeldb/invoices.go [hunk 1]
```diff
@@ -432,6 +442,11 @@ func (d *DB) AddInvoice(newInvoice *Invoice, paymentHash lntypes.Hash) (
 			return ErrDuplicateInvoice
 		}
 
+		payAddrIndex := tx.ReadWriteBucket(payAddrIndexBucket)
+		if payAddrIndex.Get(newInvoice.Terms.PaymentAddr[:]) != nil {
+			return ErrDuplicateInvoice
```
Comment (joostjager) [inline channeldb/invoices.go:447]: Distinguish between duplicate hash and addr errors?

File: channeldb/invoices.go [hunk 2]
```diff
@@ -538,7 +553,9 @@ func (d *DB) InvoicesAddedSince(sinceAddIndex uint64) ([]Invoice, error) {
 // full invoice is returned. Before setting the incoming HTLC, the values
 // SHOULD be checked to ensure the payer meets the agreed upon contractual
 // terms of the payment.
-func (d *DB) LookupInvoice(paymentHash [32]byte) (Invoice, error) {
+func (d *DB) LookupInvoice(paymentHash [32]byte,
+	payAddr *[32]byte) (Invoice, error) {
```
Comment (joostjager) [inline channeldb/invoices.go:557]: Is there a use case for specifying both?

File: channeldb/invoices.go [hunk 3]
```diff
@@ -824,7 +848,7 @@ func (d *DB) QueryInvoices(q InvoiceQuery) (InvoiceSlice, error) {
 // The update is performed inside the same database transaction that fetches the
 // invoice and is therefore atomic. The fields to update are controlled by the
 // supplied callback.
-func (d *DB) UpdateInvoice(paymentHash lntypes.Hash,
+func (d *DB) UpdateInvoice(paymentHash lntypes.Hash, payAddr *[32]byte,
```
Comment (joostjager) [inline channeldb/invoices.go:851]: As a preparation for later adding `setId`, maybe the total code delta will be smaller if an `InvoiceRef` struct is introduced containing the two (and later three) fields?

File: channeldb/invoices.go [hunk 4]
```diff
@@ -848,9 +872,16 @@ func (d *DB) UpdateInvoice(paymentHash lntypes.Hash,
 
 		// Check the invoice index to see if an invoice paying to this
 		// hash exists within the DB.
-		invoiceNum := invoiceIndex.Get(paymentHash[:])
+		var invoiceNum []byte
+		if payAddr != nil {
+			payAddrIndex := tx.ReadWriteBucket(payAddrIndexBucket)
```
Comment (joostjager) [inline channeldb/invoices.go:877]: Does this need to be a `ReadWriteBucket`? If not, maybe code from `LookupInvoice` can be reused

File: channeldb/invoices.go (hunk 2 above)
Comment (joostjager) [reply channeldb/invoices.go:557]: Or if both are specified, is a sanity check useful to see if hash is always present and pointing to the same invoice if addr is present too?

File: invoices/invoiceregistry.go [hunk 5]
```diff
@@ -560,7 +566,7 @@ func (i *InvoiceRegistry) startHtlcTimer(hash lntypes.Hash,
 // cancelSingleHtlc cancels a single accepted htlc on an invoice. It takes
 // a resolution result which will be used to notify subscribed links and
 // resolvers of the details of the htlc cancellation.
-func (i *InvoiceRegistry) cancelSingleHtlc(hash lntypes.Hash,
+func (i *InvoiceRegistry) cancelSingleHtlc(hash lntypes.Hash, payAddr *[32]byte,
```
Comment (joostjager) [inline invoices/invoiceregistry.go:569]: Yes, I think an `InvoiceRef` struct can be good for readability

File: invoices/invoiceregistry.go [hunk 6]
```diff
@@ -733,28 +739,26 @@ func (i *InvoiceRegistry) NotifyExitHopHtlc(rHash lntypes.Hash,
...
+	ctx := invoiceUpdateCtx{
 		hash:                 rHash,
 		circuitKey:           circuitKey,
 		amtPaid:              amtPaid,
 		expiry:               expiry,
 		currentHeight:        currentHeight,
 		finalCltvRejectDelta: i.cfg.FinalCltvRejectDelta,
 		customRecords:        payload.CustomRecords(),
-		mpp:                  mpp,
+		mpp:                  payload.MultiPath(),
```
Comment (joostjager) [inline invoices/invoiceregistry.go:752]: This change and the rename to `ctx` could be isolated in its own commit

File: invoices/test_utils_test.go [hunk 7]
```diff
@@ -198,14 +199,20 @@ func newTestInvoice(t *testing.T, preimage lntypes.Preimage,
 		expiry = time.Hour
 	}
 
+	var payAddr [32]byte
+	if _, err := rand.Read(payAddr[:]); err != nil {
```
Comment (joostjager) [inline invoices/test_utils_test.go:203]: It is non-deterministic, but probably that is just a theoretical problem.

File: channeldb/db.go [hunk 8]
```diff
@@ -136,6 +137,13 @@ var (
 			number:    13,
 			migration: migration13.MigrateMPP,
 		},
+		{
+			// Initialize payment address index and begin using it
+			// as the default index, falling back to payment hash
+			// index.
+			number:    14,
+			migration: mig.CreateTLB(payAddrIndexBucket),
```
Comment (joostjager) [inline channeldb/db.go:145]: Are we sure we don't want to go through the existing invoices and fill the index in the migration?

File: channeldb/invoice_test.go [hunk 9]
```diff
@@ -228,6 +280,44 @@ func TestInvoiceWorkflow(t *testing.T) {
 	}
 }
 
+// TestPayHashIndexFallback asserts that the invoice db properly prioritizes
+// lookups and updtes by payment address.
+func TestPayHashIndexFallback(t *testing.T) {
+	db, cleanUp, err := makeTestDB()
+	defer cleanUp()
+	assert.Nil(t, err)
```
Comment (joostjager) [inline channeldb/invoice_test.go:288]: Nice those assertions. Do we have enough approvals from the team to move to `testify` for `lnd`?

File: channeldb/invoice_test.go [hunk 10]
```diff
@@ -228,6 +280,44 @@ func TestInvoiceWorkflow(t *testing.T) {
...
+	invoice2, err := randInvoice(2000)
+	assert.Nil(t, err)
+
+	inv2Hash := invoice2.Terms.PaymentPreimage.Hash()
+	_, err = db.AddInvoice(invoice2, inv2Hash)
+	assert.Nil(t, err)
+
+	// Now, query using invoice 1's payment address, but invoice 2's payment
+	// hash. We expect invoice 1 to be returned since the payment hash
+	// should only be considered as a fallback.
```
Comment (joostjager) [inline channeldb/invoice_test.go:307]: This is also what the other comment about the sanity check refers to

File: channeldb/db.go (hunk 8 above)
Comment (cfromknecht) [reply channeldb/db.go:145]: we wouldn't be able to do all them because really old invoices have a pay addr of all 0's. we'd only be able to do so for invoices created post-0.8, but i'm not sure the extra work is worth it considering we have the fallback. this is definitely debatable tho

File: channeldb/invoices.go (hunk 2 above)
Comment (cfromknecht) [reply channeldb/invoices.go:557]: the reason to use both is so that we don't have to add a migration for existing invoices. with AMP the hash won't align with any particular invoice, but this wil be worked around by providing a nil hash in the `InvoiceRef`.

if we do a full migration we'd end up supporting either `(hash, nil)` or `(nil, pay_addr)`. if we don't, then we need to support `(hash, nil)`, `(nil, pay_addr)`, or (hash, pay_addr)`, where the latter has a sanity check that they point to the same invoice.

atm, there isn't really a business use case for doing the full migration, it would only serve to clean up this bit of code. this may change in the in the future tho (if we begin exposing invoice lookups by payment addr and we want to expose these slightly older invoices), but it is still possible to do this migration retroactively when that time comes. it is not terribly difficult, but it is time consuming in the critical path

File: channeldb/db.go (hunk 8 above)
Comment (joostjager) [reply channeldb/db.go:145]: Yes it would just be to not have invoices in the database that have a payment addr (post 0.8) but are not in the index.

Comment (Roasbeef) [reply channeldb/db.go:145]: Interesting migration variant. Is this needed if `CreateBucketIfNotExists` is used everywhere? Or perhaps this saves on nil checks as we can assume the bucket is always there?

File: channeldb/invoice_test.go [hunk 11]
```diff
@@ -101,31 +142,41 @@ func TestInvoiceWorkflow(t *testing.T) {
...
+	}
+	if test.queryPayAddr {
+		payAddr = &fakeInvoice.Terms.PaymentAddr
+	}
 
 	// Add the invoice to the database, this should succeed as there aren't
 	// any existing invoices within the database with the same payment
 	// hash.
-	if _, err := db.AddInvoice(fakeInvoice, paymentHash); err != nil {
+	if _, err := db.AddInvoice(fakeInvoice, invPayHash); err != nil {
```
Comment (Roasbeef) [inline channeldb/invoice_test.go:167]: How does this compare to having another method for inserting w/ a pay addr only (and no payment hash?), or adding functional arguments to allow users to specify one or both of them?

File: channeldb/invoice_test.go [hunk 12]
```diff
@@ -228,6 +280,44 @@ func TestInvoiceWorkflow(t *testing.T) {
 	}
 }
 
+// TestPayHashIndexFallback asserts that the invoice db properly prioritizes
+// lookups and updtes by payment address.
```
Comment (Roasbeef) [inline channeldb/invoice_test.go:284]: updtes -> updates

File: channeldb/invoices.go (hunk 4 above)
Comment (Roasbeef) [reply channeldb/invoices.go:877]: Functions that want a read bucket can also have a read-write bucket passed in. However in this case, we don't modify the bucket all, so we can indeed make due with a read bucket here.

File: channeldb/db.go (hunk 8 above)
Comment (cfromknecht) [reply channeldb/db.go:145]: We need to bump the version anyway, otherwise people may rollback and our invoice accounting could be incorrect. There will also be another PR adding the set_id index which will need the same.

If we can avoid it, I would prefer not using `CreateBucketIfNotExists` for static buckets that _should_ be there, and save it for buckets whose name isn't known at compile time. I added this migration to make that process easier for us going forward.
 - 1 LOC vs 4
 - fewer untested codepaths/improved coverage
 - slightly more performant
 - more clear which buckets can change and which are assumed to be part of the schema

imo these bucketing changes should be added as a migration unless it is truly optional and we want to allow users to downgrade, but i don't see that as being very likely since I can't recall the last release without a migration

File: channeldb/invoice_test.go (hunk 11 above)
Comment (cfromknecht) [reply channeldb/invoice_test.go:167]: i think we can get away with just making the payment hash a pointer, and not indexing if it's nil?

File: channeldb/invoice_test.go (hunk 10 above)
Comment (cfromknecht) [reply channeldb/invoice_test.go]: add consistency check 👍

File: channeldb/invoices.go (hunk 2 above)
Comment (cfromknecht) [reply channeldb/invoices.go]: `InvoiceRef` added, supporting `(hash, nil)` and `(hash, pay_addr)`. the final variant will be added once the related AMP changes go in

File: channeldb/db.go [hunk 13]
```diff
@@ -201,48 +201,31 @@ func (d *DB) Path() string {
...
-			return err
-		}
-		err = tx.DeleteTopLevelBucket(edgeIndexBucket)
-		if err != nil && err != kvdb.ErrBucketNotFound {
-			return err
-		}
-		err = tx.DeleteTopLevelBucket(graphMetaBucket)
-		if err != nil && err != kvdb.ErrBucketNotFound {
-			return err
+		for _, tlb := range topLevelBuckets {
```
Comment (joostjager) [inline channeldb/db.go:232]: If we're removing everything, can't it be a for each on the tx?

File: channeldb/db.go [hunk 14]
```diff
@@ -318,19 +278,12 @@ func createChannelDB(dbPath string) error {
 			return err
 		}
 
-		graphMeta, err := tx.CreateTopLevelBucket(graphMetaBucket)
-		if err != nil {
-			return err
-		}
+		graphMeta := tx.ReadWriteBucket(graphMetaBucket)
```
Comment (joostjager) [inline channeldb/db.go:290]: Go all the way and describe the initial bucket hierarchy in a tree-like datastructure?

File: channeldb/migration/create_tlb_test.go [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+import (
+	"fmt"
+	"testing"
+
+	"github.com/lightningnetwork/lnd/channeldb/kvdb"
+	"github.com/lightningnetwork/lnd/channeldb/migration"
+	"github.com/lightningnetwork/lnd/channeldb/migtest"
+)
+
+// TestCreateTLB asserts that a CreatTLB properly initializes a new top-level
```
Comment (joostjager) [inline channeldb/migration/create_tlb_test.go:12]: "CreatTLB"

File: channeldb/migration/log.go [hunk 16]
```diff
@@ -0,0 +1,12 @@
+package migration
+
+import "github.com/btcsuite/btclog"
+
+// log is a logger that is initialized as disabled.  This means the package will
+// not perform any logging by default until a logger is set.
+var log = btclog.Disabled
+
+// UseLogger uses a specified Logger to output package logging info.
+func UseLogger(logger btclog.Logger) {
```
Comment (joostjager) [inline channeldb/migration/log.go:10]: Call from main `log.go`?

File: channeldb/invoices.go [hunk 17]
```diff
@@ -142,6 +142,36 @@ const (
 	amtPaidType     tlv.Type = 13
 )
 
+// InvoiceRef is a composite identifier for invoices. Invoices can be referenced
+// by various combinations of payment hash and payment addr, in certain contexts
+// only some of these are known. An InvoiceRef and its constructors thus
+// encapsulate the valid combinations of query parameters that can be supplied
+// to LookupInvoice and UpdateInvoice.
```
Comment (joostjager) [inline channeldb/invoices.go:159]: Part of this comment is not (yet) applicable?

File: channeldb/db.go (hunk 8 above)
Comment (joostjager) [reply channeldb/db.go:145]: I tend to agree with avoiding `CreateBucketIfNotExists`. On the other hand, if we are moving away from migrations in some cases in favor of optional buckets ('on the fly migrations'), then `CreateBucketIfNotExists` is just an instance of that. If we do optional buckets for on-the-fly migrations, why not keep it for the root buckets?

One downside of this migration is that it isn't easily possible to downgrade for example to the latest tagged release during development.

File: channeldb/invoice_test.go [hunk 18]
```diff
@@ -102,32 +142,45 @@ func TestInvoiceWorkflow(t *testing.T) {
...
 	}
 
 	// Attempt to retrieve the invoice which was just added to the
 	// database. It should be found, and the invoice returned should be
 	// identical to the one created above.
 	dbInvoice, err := db.LookupInvoice(ref)
-	if err != nil {
-		t.Fatalf("unable to find invoice: %v", err)
+	if !test.queryPayAddr && !test.queryPayHash {
+		if err != ErrInvoiceNotFound {
```
Comment (joostjager) [inline channeldb/invoice_test.go:180]: If `InvoiceRef` is not initialized at all, isn't a distinct error more suitable? It is basically an invalid query

File: channeldb/invoices.go [hunk 19]
```diff
@@ -152,6 +162,13 @@ type InvoiceRef struct {
 	// currently indexed by payment hash. This value will be used as a
 	// fallback when no payment address is known.
 	payHash lntypes.Hash
+
+	// payAddr is the payment addr of the target invoice. Newer invoices
+	// (0.11 and up) are indexed by payment address in addition to payment
+	// hash, but pre 0.8 invoices do not have one at all. When this value is
+	// known it will be used as the primary identifier, falling back to
+	// payHash if no value is known.
+	payAddr *[32]byte
```
Comment (joostjager) [inline channeldb/invoices.go:171]: Not necessarily for this PR, but I can see an advantage of making an `lntypes.Hash` style datatype for payment address, especially if we are going to use it a lot

File: htlcswitch/test_utils.go [hunk 20]
```diff
@@ -2,7 +2,7 @@ package htlcswitch
 
 import (
 	"bytes"
-	"crypto/rand"
+	crand "crypto/rand"
```
Comment (joostjager) [inline htlcswitch/test_utils.go:5]: New standard for naming crypto rand?

File: channeldb/db.go (hunk 8 above)
Comment (joostjager) [reply channeldb/db.go:145]: Discussed offline. Downgrade isn't safe anyway. We probably also need to bump the version for other on-the-fly migrations

File: channeldb/invoice_test.go (hunk 18 above)
Comment (cfromknecht) [reply channeldb/invoice_test.go:180]: it is equivalent to querying with a zero hash, which we already do?

File: htlcswitch/test_utils.go (hunk 20 above)
Comment (cfromknecht) [reply htlcswitch/test_utils.go:5]: the linter kept mistaking it for `match/rand`

File: channeldb/db.go (hunk 13 above)
Comment (cfromknecht) [reply channeldb/db.go:232]: it's less atomic, but also only used for testing afaia, what would we gain?

File: channeldb/db.go (hunk 14 above)
Comment (cfromknecht) [reply channeldb/db.go:290]: seems tangential to this PR? only reason i did this for TLBs is because i would be duplicating code in adding `CreateTopLevelBucket` and `DeleteTopLevelBucket` for the pay addr index.

File: channeldb/migration/log.go (hunk 16 above)
Comment (cfromknecht) [reply channeldb/migration/log.go:13]: done!

File: channeldb/invoices.go (hunk 19 above)
Comment (cfromknecht) [reply channeldb/invoices.go:171]: yes i've been thinking maybe we should, but held off because we don't use it for anything other than comparing equality. definitely something we can look into tho, it would be nice if they were unified

File: channeldb/db.go (hunk 13 above)
Comment (joostjager) [reply channeldb/db.go:273]: Hmm so it should really be moved to a `_test.go` file then.

File: channeldb/db.go (hunk 14 above)
Comment (joostjager) [reply channeldb/db.go:326]: Yes, absolutely scope creep

File: channeldb/migration/log.go (hunk 16 above)
Comment (joostjager) [reply channeldb/migration/log.go:13]: In other areas we initialize a sub-sub-logger from the sub-logger. See for example `htlcswitch/log.go` initializing the payload logger.

File: go.sum [hunk 21]
```diff
@@ -105,6 +106,7 @@ github.com/gogo/protobuf v1.1.1 h1:72R+M5VuhED/KujmZVcIquuo8mBgX4oVda//DQb3PXo=
 github.com/gogo/protobuf v1.1.1/go.mod h1:r8qH/GZQm5c6nD/R0oafs1akxWv10x8SbQlK7atdtwQ=
 github.com/golang/glog v0.0.0-20160126235308-23def4e6c14b h1:VKtxabqXZkF25pY9ekfRL6a582T4P37/31XEstQ5p58=
 github.com/golang/glog v0.0.0-20160126235308-23def4e6c14b/go.mod h1:SBH7ygxi8pfUlaOkMMuAQtPIUF8ecWP5IEl/CR7VP2Q=
+github.com/golang/groupcache v0.0.0-20200121045136-8c9f03a8e57e h1:1r7pUrabqp18hOBcwBwiTsbnFeTZHV9eER/QT5JVZxY=
```
Comment (joostjager) [inline go.sum:109]: Does it need `tidy`?

File: channeldb/invoices.go [hunk 22]
```diff
@@ -603,18 +656,44 @@ func (d *DB) LookupInvoice(ref InvoiceRef) (Invoice, error) {
 }
 
 // fetchInvoiceNumByRef retrieve the invoice number for the provided invoice
-// reference.
-func fetchInvoiceNumByRef(invoiceIndex kvdb.ReadBucket,
+// reference. The payment address will be treated as the primary key, falling
+// back to the payment hash if nothing is found for the payment address. An
+// errors is returned if the invoice is not found.
```
Comment (joostjager) [inline channeldb/invoices.go:661]: an errors

File: channeldb/invoice_test.go (hunk 18 above)
Comment (joostjager) [reply channeldb/invoice_test.go:180]: Hm true. Maybe the optional hash will arrive with amp then.

File: go.sum (hunk 21 above)
Comment (cfromknecht) [reply go.sum:109]: fixed in #4315

**Comment by Roasbeef** [comment]
Test failure on travis: 
```
--- FAIL: TestChannelLinkBatchPreimageWrite (7.15s)
    --- FAIL: TestChannelLinkBatchPreimageWrite/flush_on_disconnect (6.20s)
        link_test.go:4746: unable to find preimages: expected to find witness: true, got false for hash=7888bd181c582842696baa9d3a55f8e324637da1c91dc89690aa71e3a04c91d0
```

Not sure if related.

**Comment by Roasbeef** [comment]
New trigger in the error log filtration for the itests: 
```
PASS
ok  	github.com/lightningnetwork/lnd/lntest/itest	990.562s
lntest/itest/log_check_errors.sh
<time> [ERR] NTFN: unable to get hash from block with height 790
```

File: channeldb/db.go (hunk 13 above)
Comment (Roasbeef) [reply channeldb/db.go:273]: Yeah it's something left over from super early development, when the easiest thing to do was just delete the entire database and start over lol.

File: channeldb/invoices.go [hunk 23]
```diff
@@ -142,6 +142,32 @@ const (
 	amtPaidType     tlv.Type = 13
 )
 
+// InvoiceRef is an identifier for invoices supporting queries by payment hash.
```
Comment (Roasbeef) [inline channeldb/invoices.go]: me gusta

File: invoices/invoiceregistry.go [hunk 24]
```diff
@@ -962,15 +966,16 @@ func (i *InvoiceRegistry) SettleHodlInvoice(preimage lntypes.Preimage) error {
...
+	invoice, err := i.cdb.UpdateInvoice(invoiceRef, updateInvoice)
 	if err != nil {
 		log.Errorf("SettleHodlInvoice with preimage %v: %v",
 			preimage, err)
 
 		return err
 	}
 
-	log.Debugf("Invoice(%v): settled with preimage %v", hash,
+	log.Debugf("Invoice%v: settled with preimage %v", invoiceRef,
```
Comment (Roasbeef) [inline invoices/invoiceregistry.go:978]: Re-add the parens here? Otherwise everything will be smushed together. 

(this log line)

File: channeldb/invoice_test.go [hunk 25]
```diff
@@ -991,9 +1108,17 @@ func TestCustomRecords(t *testing.T) {
...
 	// payment addr.
 	refByHash := InvoiceRefByHash(payHash)
 	assert.Equal(t, payHash, refByHash.PayHash())
+	assert.Equal(t, (*[32]byte)(nil), refByHash.PayAddr())
+
+	// An InvoiceRef by hash and addr should return the payment hash and
+	// payment addr passed to the constructor.
+	refByHashAndAddr := InvoiceRefByHashAndAddr(payHash, payAddr)
+	assert.Equal(t, payHash, refByHashAndAddr.PayHash())
+	assert.Equal(t, &payAddr, refByHashAndAddr.PayAddr())
```
Comment (Roasbeef) [inline channeldb/invoice_test.go:1123]: Not gonna lie, this `assert` package is pretty nifty lol.

File: channeldb/invoices.go [hunk 26]
```diff
@@ -158,13 +179,39 @@ func InvoiceRefByHash(payHash lntypes.Hash) InvoiceRef {
...
 func (r InvoiceRef) PayHash() lntypes.Hash {
 	return r.payHash
 }
 
+// PayAddr returns the optional payment address of the target invoice.
+//
+// NOTE: This value may be nil.
+func (r InvoiceRef) PayAddr() *[32]byte {
+	if r.payAddr != nil {
+		addr := *r.payAddr
```
Comment (Roasbeef) [inline channeldb/invoices.go:204]: 👍

File: htlcswitch/test_utils.go (hunk 20 above)
Comment (Roasbeef) [reply htlcswitch/test_utils.go:5]: They should totally have distinct packages name imo...

File: invoices/invoiceregistry.go (hunk 24 above)
Comment (cfromknecht) [reply invoices/invoiceregistry.go:978]: the `String` method now adds the parens

Comment (cfromknecht) [reply invoices/invoiceregistry.go:978]: did a scan to convert any other log statements to print using the invoice ref so everything is consistent

**Comment by githorray** [comment]
Getting an error after this merge when trying to build lnd.

```
$ make install tags="experimental signrpc walletrpc chainrpc invoicesrpc"

github.com/lightningnetwork/lnd/channeldb
# github.com/lightningnetwork/lnd/channeldb
channeldb/invoices.go:657:54: undefined: kvdb.ReadBucket
Makefile:115: recipe for target 'build' failed
make: *** [build] Error 2```
